server/pages/serviceWorker.js

```diff
@@ -0,0 +1,141 @@
+				// fetch handler as responses directly, without consulting the network (i.e. cache-first).
+				// If we were to get back a response from the HTTP browser cache for this precaching request
+				// then that stale response would be used indefinitely, or at least until the next time
+				// the service worker script changes triggering the install flow.
+				url.search += ( url.search ? '&' : '?' ) + 'cache-bust=' + now;
+
+				// It's very important to use {mode: 'no-cors'} if there is any chance that
+				// the resources being fetched are served off of a server that doesn't support
+				// CORS (http://en.wikipedia.org/wiki/Cross-origin_resource_sharing).
+				// In this example, www.chromium.org doesn't support CORS, and the fetch()
+				// would fail if the default mode of 'cors' was used for the fetch() request.
+				// The drawback of hardcoding {mode: 'no-cors'} is that the response from all
+				// cross-origin hosts will always be opaque
+				// (https://slightlyoff.github.io/ServiceWorker/spec/service_worker/index.html#cross-origin-resources)
+				// and it is not possible to determine whether an opaque response represents a success or failure
+				// (https://github.com/whatwg/fetch/issues/14).
+				return fetch( new Request( url, { mode: 'no-cors' } ) ).then( function( response ) {
+					if ( response.status >= 400 ) {
+						throw new Error( 'request for ' + urlToPrefetch +
+							' failed with status ' + response.statusText);
+					}
+
+					// Use the original URL without the cache-busting parameter as the key for cache.put().
+					return cache.put( urlToPrefetch, response );
+				} ).catch( function( error ) {
+					console.error( 'Not caching ' + urlToPrefetch + ' due to ' + error );
+				} );
+			} );
+
+			return Promise.all( cachePromises ).then( function() {
+				console.log( 'Pre-fetching complete.' );
+			} );
+		} ).catch( function( error ) {
+			console.error( 'Pre-fetching failed:', error );
+		} )
+	);
+} );
+
+self.addEventListener( 'activate' , function( event ) {
+	// Delete all caches that aren't named in CURRENT_CACHES.
+	// While there is only one cache in this example, the same logic will handle the case where
+	// there are multiple versioned caches.
+	var expectedCacheNames = Object.keys( CURRENT_CACHES ).map( function( key ) {
+		return CURRENT_CACHES[key];
+	} );
+
+	event.waitUntil(
+		caches.keys().then( function( cacheNames ) {
+			return Promise.all(
+				cacheNames.map( function( cacheName ) {
+					if ( expectedCacheNames.indexOf( cacheName ) === -1 ) {
+						// If this cache name isn't present in the array of "expected" cache names, then delete it.
+						console.log( 'Deleting out of date cache:', cacheName );
+						return caches.delete( cacheName );
+					}
+				} )
+			);
+		} )
+	);
+} );
+
+self.addEventListener( 'fetch', function( event ) {
+	console.log( 'Handling fetch event for', event.request.url );
+
+	event.respondWith(
```

Comment:
This is a simple cache-first policy. The more complex policy of cache-first-with-background-update could be implemented as follows:

```js
caches.match( event.request ).then( function( response ) {
			var fetchFromNetwork = fetch( event.request.clone()).then( function( response ) {
				console.log( 'Response from network is:', response );
				return caches.open().then(function(cache) {
					return cache.put(event.request, response.clone()).then(function() { return response.clone(); });
				});
			} ).catch( function( error ) {
				console.error( 'Fetching failed:', error );
				throw error;
			} );
			return response ? Promise.resolve(response.clone()) || fetchFromNetwork;
		} )
```